bgpd/bgp_pbr.c

```diff
@@ -0,0 +1,1050 @@
+				 __func__, afi, family2afi(dst->family));
+			return -1;
+		}
+	}
+	return 0;
+}
+
+static int bgp_pbr_match_counter_unique;
+static int bgp_pbr_match_entry_counter_unique;
+static int bgp_pbr_action_counter_unique;
+
+static void *bgp_pbr_match_alloc_intern(void *arg)
+{
+	struct bgp_pbr_match *bpm, *new;
+
+	bpm = (struct bgp_pbr_match *)arg;
+
+	new = XCALLOC(MTYPE_TMP, sizeof(*new));
+
+	memcpy(new, bpm, sizeof(*bpm));
+
+	return new;
+}
+
+static void *bgp_pbr_action_alloc_intern(void *arg)
+{
+	struct bgp_pbr_action *bpa, *new;
+
+	bpa = (struct bgp_pbr_action *)arg;
+
+	new = XCALLOC(MTYPE_TMP, sizeof(*new));
+
+	memcpy(new, bpa, sizeof(*bpa));
+
+	return new;
+}
+
+static void *bgp_pbr_match_entry_alloc_intern(void *arg)
+{
+	struct bgp_pbr_match_entry *bpme, *new;
+
+	bpme = (struct bgp_pbr_match_entry *)arg;
+
+	new = XCALLOC(MTYPE_TMP, sizeof(*new));
+
+	memcpy(new, bpme, sizeof(*bpme));
+
+	return new;
+}
+
+uint32_t bgp_pbr_match_hash_key(void *arg)
+{
+	struct bgp_pbr_match *pbm = (struct bgp_pbr_match *)arg;
+	uint32_t key;
+
+	key = jhash_1word(pbm->vrf_id, 0x4312abde);
+	key = jhash_1word(pbm->flags, key);
+	return jhash_1word(pbm->type, key);
+}
+
+int bgp_pbr_match_hash_equal(const void *arg1, const void *arg2)
+{
+	const struct bgp_pbr_match *r1, *r2;
+
+	r1 = (const struct bgp_pbr_match *)arg1;
+	r2 = (const struct bgp_pbr_match *)arg2;
+
+	if (r1->vrf_id != r2->vrf_id)
+		return 0;
+
+	if (r1->type != r2->type)
+		return 0;
+
+	if (r1->flags != r2->flags)
+		return 0;
+
+	if (r1->action != r2->action)
+		return 0;
+
+	/* unique value is self calculated
+	 */
+
+	/* ipset name is self calculated
+	 */
+	return 1;
+}
+
+uint32_t bgp_pbr_match_entry_hash_key(void *arg)
+{
+	struct bgp_pbr_match_entry *pbme;
+	uint32_t key;
+
+	pbme = (struct bgp_pbr_match_entry *)arg;
+	key = prefix_hash_key(&pbme->src);
+	key = jhash_1word(prefix_hash_key(&pbme->dst), key);
+
+	return key;
+}
+
+int bgp_pbr_match_entry_hash_equal(const void *arg1, const void *arg2)
+{
+	const struct bgp_pbr_match_entry *r1, *r2;
+
+	r1 = (const struct bgp_pbr_match_entry *)arg1;
+	r2 = (const struct bgp_pbr_match_entry *)arg2;
+
+	/* on updates, comparing
+	 * backpointer is not necessary
+	 */
+
+	/* unique value is self calculated
+	 */
+
+	/* rate is ignored for now
+	 */
+
+	if (!prefix_same(&r1->src, &r2->src))
+		return 0;
+
+	if (!prefix_same(&r1->dst, &r2->dst))
+		return 0;
+
+	return 1;
+}
+
+uint32_t bgp_pbr_action_hash_key(void *arg)
+{
+	struct bgp_pbr_action *pbra;
+	uint32_t key;
+
+	pbra = (struct bgp_pbr_action *)arg;
+	key = jhash_1word(pbra->table_id, 0x4312abde);
+	key = jhash_1word(pbra->fwmark, key);
+	return key;
+}
+
+int bgp_pbr_action_hash_equal(const void *arg1, const void *arg2)
+{
+	const struct bgp_pbr_action *r1, *r2;
+
+	r1 = (const struct bgp_pbr_action *)arg1;
+	r2 = (const struct bgp_pbr_action *)arg2;
+
+	/* unique value is self calculated
+	 */
+	/* table and fwmark is self calculated
+	 */
```

Comment:
Why are these two comments instead of one

Reply:
gathered in one